File: nonet/src/main/java/com/keiferstone/nonet/NoNet.java
package com.keiferstone.nonet;

import android.content.Context;

/**
 * An Android library for monitoring network connectivity.
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public final class NoNet {
    private static NoNet instance = null;

    private Configuration configuration;
    private Monitor monitor;

    private NoNet() {}

    /**
     * Set the global configuration.
     *
     * @return A {@link Configuration.Builder} that is automatically applied to the {@link NoNet} instance.
     */
    public static Configuration.Builder configure() {
        instantiate();
        Configuration.Builder builder = new Configuration.Builder();
        instance.configuration = builder.configuration;
        return builder;
    }

    /**
     * Start monitoring network connectivity.
     *
     * @param context Context for listening to connectivity events and displaying notifications.
     *                Must be an instance of {@link android.app.Activity} for
     *                {@link Monitor.Builder#snackbar()} to work. Don't forget to call
     *                {@link #stopMonitoring()} to stop monitoring.
     *
     * @return A {@link Monitor.Builder}.
     */
    public static Monitor.Builder monitor(Context context) {
        instantiate();
        stopMonitoring();
        Monitor.Builder builder = new Monitor.Builder(context);
        if (instance.configuration != null) {
            builder.configure(instance.configuration);
        }
        instance.monitor = builder.monitor;
        return builder;
    }

    /**
     * Make a single check for network connectivity.
     *
     * @param context Context for displaying notifications. Must be an instance of
     *                {@link android.app.Activity} for {@link Check.Builder#snackbar()}
     *                to work.
     *
     * @return A {@link Check.Builder}.
     */
    public static Check.Builder check(Context context) {
        instantiate();
        Check.Builder builder = new Check.Builder(context);
        if (instance.configuration != null) {
            builder.configure(instance.configuration);
        }
        return builder;
    }

    /**
     * Stop monitoring network connectivity.
     */
    public static void stopMonitoring() {
        instantiate();
        if (instance.monitor != null) {
            instance.monitor.stop();
            instance.monitor = null;
        }
    }

    private static void instantiate() {
        if (instance == null) {
            instance = new NoNet();
        }
    }
}

File: nonet/src/main/java/com/keiferstone/nonet/NoNet.kt
package com.keiferstone.nonet

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.lifecycle.*
import okhttp3.*
import java.io.IOException
import java.util.concurrent.TimeUnit


@Suppress("unused")
object NoNet {
    private lateinit var connectivityManager: ConnectivityManager
    private lateinit var networkCallback: ConnectivityManager.NetworkCallback

    fun init(context: Context) {
        connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    }

    /**
     * Check the system [ConnectivityManager] to determine if we have an active network connection.
     */
    fun hasConnection(): Boolean {
        return connectivityManager.activeNetworkInfo?.isConnected == true
    }

    /**
     * Check if we have a network connection.
     *
     * @param config An optional custom configuration
     */
    suspend fun isConnected(config: Config = Config()): Boolean {
        return try {
            val request = Request.Builder()
                    .url(config.url)
                    .build()
            OkHttpClient.Builder()
                    .callTimeout(config.timeout, TimeUnit.SECONDS)
                    .build()
                    .newCall(request)
                    .execute().use { response ->
                response.isSuccessful
            }
        } catch (e: IOException) {
            Log.e(NoNet.javaClass.name, "Connection check failed", e)
            false
        }
    }

    /**
     * Check if we have a network connection.
     *
     * @param config An optional custom configuration
     */
    fun isConnected(config: Config = Config(), callback: (Boolean) -> Unit) {
        val request = Request.Builder()
                .url(config.url)
                .build()
        OkHttpClient.Builder()
                .callTimeout(config.timeout, TimeUnit.SECONDS)
                .build()
                .newCall(request)
                .enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                callback(response.isSuccessful)
            }

            override fun onFailure(call: Call, e: IOException) {
                Log.e(NoNet.javaClass.name, "Connection check failed", e)
                callback(false)
            }
        })
    }


    /**
     * Monitor network connectivity as long as the provided [lifecycleOwner] is resumed.
     *
     * @param lifecycleOwner The [LifecycleOwner] to attach to
     * @param config An optional custom configuration
     */
    fun monitorConnection(lifecycleOwner: LifecycleOwner, config: Config = Config(), callback: (Boolean) -> Unit) {
        lifecycleOwner.lifecycle.addObserver(ConnectionObserver(config, lifecycleOwner, callback))
    }

    private class ConnectionObserver(private val config: Config,
                                     private val lifecycleOwner: LifecycleOwner,
                                     private val callback: (Boolean) -> Unit) : LifecycleObserver {
        private val monitorHandler = Handler()
        private val runnable = object : Runnable {
            private val callbackHandler = Handler(Looper.getMainLooper())

            override fun run() {
                isConnected(config) {
                    if (config.callbackOnlyIfChanged) {
                        if (isConnected != it) callbackHandler.post { callback(it) }
                    } else callbackHandler.post { callback(it) }
                    isConnected = it
                    val pollInterval =
                            if (it) config.connectedPollInterval
                            else config.disconnectedPollInterval
                    if (pollInterval > 0) {
                        monitorHandler.postDelayed(this, pollInterval * 1000)
                    }
                }
            }
        }
        private val networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                Log.d(NoNet.javaClass.name, "Network available")
                monitorHandler.removeCallbacks(runnable)
                monitorHandler.post(runnable)
            }

            override fun onLost(network: Network?) {
                Log.d(NoNet.javaClass.name, "Network lost")
                monitorHandler.removeCallbacks(runnable)
                monitorHandler.post(runnable)
            }

            override fun onUnavailable() {
                Log.d(NoNet.javaClass.name, "Network unavailable")
                monitorHandler.removeCallbacks(runnable)
                monitorHandler.post(runnable)
            }
        }
        private val request =
                NetworkRequest.Builder()
                        .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
                        .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
                        .build()
        private var isConnected: Boolean? = null

        @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
        fun startMonitoring() {
            Log.d(NoNet.javaClass.name, "Starting connection monitoring")
            connectivityManager.registerNetworkCallback(request, networkCallback)
            monitorHandler.post(runnable)
        }

        @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
        fun stopMonitoring() {
            Log.d(NoNet.javaClass.name, "Stopping connection monitoring")
            connectivityManager.unregisterNetworkCallback(networkCallback)
            monitorHandler.removeCallbacks(runnable)
        }

        @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        fun destroy() {
            Log.d(NoNet.javaClass.name, "Destroying connection observer")
            lifecycleOwner.lifecycle.removeObserver(this)
        }
    }
}